### PR TITLE
[tools/setup/setup.sh] Nightly cleanup

### PR DESCRIPTION
**Pull request by SamuelMarks**
# Description

[tools/setup/setup.sh] Cleanup redundant and unnecessary nightly dependencies ; use `--no-deps` ; minor other cleanup / DRYness

# Tests

CI and manual:
```sh
$ bash ./dependencies/scripts/docker_build_dependency_image.sh DEVICE='tpu' MODE='nightly'
```
```sh
$ bash ./dependencies/scripts/docker_build_dependency_image.sh DEVICE='tpu' MODE='stable'
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tools/setup/setup.sh [hunk 1]
```diff
@@ -261,42 +240,43 @@ elif [[ $MODE == "nightly" ]]; then
...
-        python3 -m uv pip install 'git+https://github.com/google/tunix.git'
-        echo "Installing jax-nightly, jaxlib-nightly"
-        # Install jax-nightly
-        python3 -m uv pip install --pre -U jax -i https://us-python.pkg.dev/ml-oss-artifacts-published/jax/simple/
-        # Install jaxlib-nightly
-        python3 -m uv pip install --pre -U jaxlib -i https://us-python.pkg.dev/ml-oss-artifacts-published/jax/simple/
+        if [[ -n "$JAX_VERSION" ]]; then
+            echo "Installing jax-nightly, jaxlib-nightly ${JAX_VERSION}"
+            python3 -m uv pip install --no-deps -U --pre jax=="${JAX_VERSION}" jaxlib=="${JAX_VERSION}" \
+                -i 'https://us-python.pkg.dev/ml-oss-artifacts-published/jax/simple/'
```
Comment (bvandermoon) [inline tools/setup/setup.sh:229]: If JAX version is used, we will also need to use a libtpu installation from the same day so that it's compatible

**Review by bvandermoon — COMMENTED** [review]
Thanks Samuel. Could you please also run a training for both of the `MODE` options being updated here? As well as with/without explicitly setting `JAX_VERSION`?

File: tools/setup/setup.sh (hunk 1 above)
Comment (SamuelMarks) [reply tools/setup/setup.sh:229]: I can do that - hey do you have a working link https://us-python.pkg.dev/ml-oss-artifacts-published/jax-public-nightly-artifacts-registry/simple from https://docs.jax.dev/en/latest/installation.html#jax-nightly-installation

(I'm cross-referencing with https://storage.googleapis.com/jax-releases/libtpu_releases.html)

EDIT: Posted here also https://github.com/jax-ml/jax/issues/33341

**Comment by bvandermoon** [comment]
> Thanks Samuel. Could you please also run a training for both of the `MODE` options being updated here? As well as with/without explicitly setting `JAX_VERSION`?

It would also be good to `pip show` that the versions are what you would expect for each build

File: tools/setup/setup.sh [hunk 2]
```diff
@@ -220,36 +178,34 @@ fi
...
     if [[ $DEVICE == "tpu" ]]; then
-        # TODO: Once tunix has support for GPUs, move it from here to requirements.txt
-        echo "Installing google-tunix for stable TPU environment"
-        python3 -m uv pip install 'google-tunix>=0.1.2'
         echo "Installing stable jax, jaxlib for tpu"
         if [[ -n "$JAX_VERSION" ]]; then
             echo "Installing stable jax, jaxlib, libtpu version ${JAX_VERSION}"
-            python3 -m uv pip install -U jax[tpu]==${JAX_VERSION} -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+            python3 -m uv pip install -U jax[tpu]=="${JAX_VERSION}" \
+                -f 'https://storage.googleapis.com/jax-releases/libtpu_releases.html'
```
Comment (bvandermoon) [inline tools/setup/setup.sh:187]: Looks like this should be with `--no-deps`

File: tools/setup/setup.sh [hunk 3]
```diff
@@ -220,36 +178,34 @@ fi
...
         fi
         if [[ -n "$LIBTPU_GCS_PATH" ]]; then
             # Install custom libtpu
             echo "Installing libtpu.so from $LIBTPU_GCS_PATH to $libtpu_path"
             # Install required dependency
-            python3 -m uv pip install -U crcmod
+            python3 -m uv pip install -U 'crcmod'
```
Comment (bvandermoon) [inline tools/setup/setup.sh]: Should this be with `--no-deps`?

File: tools/setup/setup.sh [hunk 4]
```diff
@@ -261,42 +217,43 @@ elif [[ $MODE == "nightly" ]]; then
...
             # Copy libtpu.so from GCS path
             gsutil cp "$LIBTPU_GCS_PATH" "$libtpu_path"
         else
             # Install libtpu-nightly
             echo "Installing libtpu-nightly"
-            python3 -m uv pip install -U --pre libtpu -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+            python3 -m uv pip install -U --pre 'libtpu' -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
```
Comment (bvandermoon) [inline tools/setup/setup.sh]: This should be with `--no-deps`

File: tools/setup/setup.sh (hunk 3 above)
Comment (SurbhiJainUSC) [reply tools/setup/setup.sh]: Looks like this is needed only when `device=tpu` and `LIBTPU_GCS_PATH` is set. Can we remove all occurrences of `pip install -U crcmod` and move it top? It will lead to only one-time installation of this package.

Comment (SamuelMarks) [reply tools/setup/setup.sh]: Guarded by `if [[ -n "$LIBTPU_GCS_PATH" ]]` I assume; as it's not needed in other branches

File: tools/setup/setup.sh (hunk 1 above)
Comment (bvandermoon) [reply tools/setup/setup.sh:229]: I think the second link you have (https://storage.googleapis.com/jax-releases/libtpu_releases.html) is the recommended way to install libtpu nightly at this point

File: tools/setup/setup.sh [hunk 5]
```diff
@@ -220,83 +180,73 @@ fi
...
 elif [[ $MODE == "nightly" ]]; then
 # Nightly mode
 
     # Uninstall existing jax, jaxlib and  libtpu-nightly
-    python3 -m uv pip show jax && python3 -m uv pip uninstall jax
-    python3 -m uv pip show jaxlib && python3 -m uv pip uninstall jaxlib
-    python3 -m uv pip show libtpu-nightly && python3 -m uv pip uninstall libtpu-nightly
+    python3 -m uv pip show 'jax' && python3 -m uv pip uninstall 'jax'
+    python3 -m uv pip show 'jaxlib' && python3 -m uv pip uninstall 'jaxlib'
+    python3 -m uv pip show 'libtpu-nightly' && python3 -m uv pip uninstall 'libtpu-nightly'
```
Comment (bvandermoon) [inline tools/setup/setup.sh:211]: Does uninstalling update any transitive dependencies?

Comment (SamuelMarks) [reply tools/setup/setup.sh:211]: This is identical to previous functionality. `uv pip uninstall` does not have options to remove or not remove transitive dependencies. I don't believe it does, see for example:
```sh
(maxtext-uv-3-12) $ uv pip uninstall fastapi
Using Python 3.12.12 environment at: venvs/maxtext-uv-3-12
Uninstalled 1 package in 56ms
 - fastapi==0.119.1
(maxtext-uv-3-12) $ uv pip install fastapi
Using Python 3.12.12 environment at: venvs/maxtext-uv-3-12
Resolved 11 packages in 450ms
Prepared 2 packages in 32ms
Installed 2 packages in 4ms
 + annotated-doc==0.0.4
 + fastapi==0.121.2
```

File: tools/setup/setup.sh (hunk 4 above)
Comment (SamuelMarks) [reply tools/setup/setup.sh]: Are you sure you want that, even for `libtpu`?

File: tools/setup/setup.sh [hunk 6]
```diff
@@ -147,69 +149,27 @@ if [[ $DEVICE == "tpu" ]]; then
...
-      'gpu') python3 -m uv pip install -e .[cuda12] --no-deps --resolution=lowest ;;
-      'tpu') python3 -m uv pip install -e .[tpu] --no-deps --resolution=lowest ;;
+      'gpu') python3 -m uv pip install -e .'[cuda12]' --no-deps --resolution='lowest' ;;
+      'tpu') python3 -m uv pip install -e .'[tpu]' --no-deps --resolution='lowest' ;;
       *)
           >&2 printf 'Unsupported device\n'
           exit 6
           ;;
   esac
-  python3 -m uv pip install --resolution=lowest -r 'src/install_maxtext_extra_deps/extra_deps_from_github.txt'
```
Comment (SurbhiJainUSC) [inline tools/setup/setup.sh:212]: Why are we install deps from Github again? It looks redundant to me. We have already installed them at line 159, right?

File: tools/setup/setup.sh (hunk 2 above)
Comment (SurbhiJainUSC) [reply tools/setup/setup.sh:187]: If `tpu-requirements.txt` were generated with JAX==0.8.0 (say) and then if we install `JAX_VERSION=0.7.2` without deps installation, then we are not even sure if it will work or not. There might be some dependency conflicts. How do we ensure that everything works?

File: tools/setup/setup.sh (hunk 6 above)
Comment (SamuelMarks) [reply tools/setup/setup.sh:212]: Yes but this happens after. So caching in Docker is setup such that the first run doesn't have our package at all so it never hits that branch. Installing from `extra_deps_from_github.txt` in stage 1 is correct, then in stage 2 it may install from there again but will be cached so at most will override conflicts from `pip install` if `--no-deps` is ever removed. The duplication is intended.

**Comment by SurbhiJainUSC** [comment]
> > Thanks Samuel. Could you please also run a training for both of the `MODE` options being updated here? As well as with/without explicitly setting `JAX_VERSION`?
> 
> It would also be good to `pip show` that the versions are what you would expect for each build

@SamuelMarks - Can you also test the setup when `MODE` and `JAX_VERSION` are specified?

File: tools/setup/setup.sh [hunk 7]
```diff
@@ -223,75 +183,70 @@ if [[ "$MODE" == "stable" || ! -v MODE ]]; then
...
             echo "Installing libtpu.so from $LIBTPU_GCS_PATH to $libtpu_path"
-            # Install required dependency
-            python3 -m uv pip install -U crcmod
             # Copy libtpu.so from GCS path
             gsutil cp "$LIBTPU_GCS_PATH" "$libtpu_path"
         else
             # Install libtpu-nightly
             echo "Installing libtpu-nightly"
-            python3 -m uv pip install -U --pre libtpu -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
+            python3 -m uv pip install -U --pre 'libtpu' \
```
Comment (SurbhiJainUSC) [inline tools/setup/setup.sh:246]: This should be with `--no-deps` right?

File: tools/setup/setup.sh [hunk 8]
```diff
@@ -128,15 +128,17 @@ if [[ -n $JAX_VERSION && ! ($MODE == "stable" || -z $MODE || ($MODE == "nightly"
     exit 1
 fi
 
+if [[ -n "$LIBTPU_GCS_PATH" ]]; then
+    python3 -m uv pip install -U 'crcmod'
```
Comment (kanglant) [inline tools/setup/setup.sh:132]: Can you check if this dep is necessary? If so, please add a comment to explain why.

File: tools/setup/setup.sh [hunk 9]
```diff
@@ -147,69 +149,27 @@ if [[ $DEVICE == "tpu" ]]; then
...
+printf 'Installing "%s" with pinned commits.\n' "${dep_name##*/}"
+python3 -m uv pip install --resolution=lowest -r "$dep_name" \
+                                              -r 'src/install_maxtext_extra_deps/extra_deps_from_github.txt'
+
 # Install maxtext package
 if [ -f 'pyproject.toml' ]; then
   case "$DEVICE" in
-      'gpu') python3 -m uv pip install -e .[cuda12] --no-deps --resolution=lowest ;;
-      'tpu') python3 -m uv pip install -e .[tpu] --no-deps --resolution=lowest ;;
+      'gpu') python3 -m uv pip install -e .'[cuda12]' --no-deps --resolution='lowest' ;;
```
Comment (kanglant) [inline tools/setup/setup.sh:165]: We'll need to add DEVICE=cuda12 and DEVICE=cuda13 values to support both CUDA versions. This is just a reminder for future work, not a change required for this PR.

File: tools/setup/setup.sh (hunk 6 above)
Comment (kanglant) [reply tools/setup/setup.sh:212]: Do you mean that the `--no-deps` will be removed from the package installation commands in the future, so the second github deps installation command here is to act as a safeguard to override any conflicting package versions?

File: tools/setup/setup.sh (hunk 2 above)
Comment (kanglant) [reply tools/setup/setup.sh:187]: As we discussed, if users choose to install a JAX version different from the one pinned by MaxText, we must present a warning. We cannot guarantee consistent dependency environments while offering this level of JAX version flexibility (maintaining lock files for every JAX/libtpu combination is not feasible).

Since JAX==0.8.0 is already included in tpu-requirements.txt, you don't need to install it again here. I recommend the following steps in this block:
1. Define Variables: Introduce variables here for the pinned JAX version (and optionally the corresponding libtpu version) used during requirement generation.
2. Conditional Installation: Only execute the installation logic if the requested JAX version differs from the pinned one.
3. Warning/Flag Choice: When installing the different version, you can either:
 * Install without `--no-deps` but issue a strong warning that the resulting environment may be inconsistent.
 * Install with `--no-deps` but issue a strong warning that the dependency versions may be incompatible with the new JAX version.

File: tools/setup/setup.sh [hunk 10]
```diff
@@ -223,75 +183,70 @@ if [[ "$MODE" == "stable" || ! -v MODE ]]; then
...
 elif [[ $MODE == "nightly" ]]; then
 # Nightly mode
 
     # Uninstall existing jax, jaxlib and  libtpu-nightly
-    python3 -m uv pip show jax && python3 -m uv pip uninstall jax
-    python3 -m uv pip show jaxlib && python3 -m uv pip uninstall jaxlib
-    python3 -m uv pip show libtpu-nightly && python3 -m uv pip uninstall libtpu-nightly
+    python3 -m uv pip show 'jax' && python3 -m uv pip uninstall 'jax'
+    python3 -m uv pip show 'jaxlib' && python3 -m uv pip uninstall 'jaxlib'
+    python3 -m uv pip show 'libtpu-nightly' && python3 -m uv pip uninstall 'libtpu-nightly'
```
Comment (kanglant) [inline tools/setup/setup.sh:211]: Given that libtpu has nightly versions like `libtpu==0.0.32.dev20251210` available in the [index](https://storage.googleapis.com/jax-releases/libtpu_releases.html) , should we also uninstall existing `libtpu` package?

File: tools/setup/setup.sh (hunk 7 above)
Comment (kanglant) [reply tools/setup/setup.sh:246]: Please explain. We use the `--no-deps` flag for JAX nightly artifacts, but not for the `libtpu` nightly. Could you explain the reasoning for this difference? Add a short comment in the script would be helpful for users.

File: tools/setup/setup.sh (hunk 8 above)
Comment (SamuelMarks) [reply tools/setup/setup.sh:132]: Pre-existing dep. If it works without this you can remove this dep.

File: tools/setup/setup.sh (hunk 7 above)
Comment (SamuelMarks) [reply tools/setup/setup.sh:246]: No because if libtpu depends on something then that version needs to be what it provides. Otherwise we'll be reporting bugs to libtpu that are not their bug but ours.

**Comment by SurbhiJainUSC** [comment]
Comments will be addressed as part of https://github.com/AI-Hypercomputer/maxtext/pull/2833. Closing this PR.
